File: app/presenters/sublink_presenter.rb
class SublinkPresenter
  attr_reader :object, :options

  PARENT_TITLE = 'All'.freeze
  MORE_COLLECTIONS_SUBLINK = {
    title: 'More',
    url: '/collections/p/1',
    icon: 'arrow-right',
    selected: false
  }.freeze

  delegate :cid, :is_main?, :is_primary?, :has_parent?, :siblings, :children,
           :root_url, to: :object

  def initialize(object, options: {})
    @object = object
    @options = options
  end

  def sublinks
    if is_main?
      return Collection.
              primary.
              ranked.
              map{ |collection| to_sublink(collection) }.
              flatten.
              append(MORE_COLLECTIONS_SUBLINK)
    end

    return [] if !is_primary? && !has_parent?
    return [] if is_primary? && children.empty?

    links = has_parent? ? siblings : children
    links.
      map { |link| to_sublink(link, link.cid === cid) }.
      sort_by { |link| link[:title] }.
      prepend({
        title: PARENT_TITLE,
        url: "#{root_url}/p/1",
        selected: is_primary?
      })
  end

  private

  def to_sublink(collection, selected = false)
    {
      title: collection.mini_title,
      url: collection.first_page_slug,
      selected: selected
    }
  end
end

File: src/presenters/sublinkPresenter.js
import Collection from '#src/models/collection.js';
import settings from '#src/config/settings.js';

export default class SublinkPresenter {
  constructor(object, options = {}) {
    this.object = object;
    this.options = options;
  }

  get sublinks() {
    if (this.object.isMain) {
      return Collection.scope('primary', 'byRanking')
        .filter(collection => !collection.isUpdateLogs)
        .map(collection => this.toSublink(collection))
        .flat()
        .concat([settings.sublinks.moreCollections]);
    }

    if (!this.object.isPrimary && !this.object.hasParent) return [];

    if (this.object.isPrimary && this.object.children.length === 0) return [];

    let links = this.object.hasParent
      ? this.object.siblings
      : this.object.children;

    console;

    links = links
      .map(link => this.toSublink(link, link.id === this.object.id))
      .sort((a, b) => a.title.localeCompare(b.title));

    links.unshift({
      title: settings.sublinks.parentTitle,
      url: `${this.object.rootUrl}/p/1`,
      selected: this.object.isPrimary,
    });

    return links;
  }

  toSublink(collection, selected = false) {
    return {
      title: collection.miniTitle,
      url: collection.firstPageSlug,
      selected,
    };
  }
}
